### PR TITLE
Onboard ESLint and prettier

### PR DESCRIPTION
**Pull request by anywherepilot**
Remove TSLint, since it was deprecated last year.

### CODE REVIEW

**Comment by anywherepilot** [comment]
Please review each commit by itself, especially the one called "Ignore most and fix some ESLint errors".

**Comment by anywherepilot** [comment]
This addresses #770

File: gulpfile.js [hunk 1]
```diff
@@ -1,97 +1,84 @@
...
-  rimraf('dist', done);
+gulp.task("clean", function (done) {
+  rimraf("dist", done);
 });
 
-gulp.task('build-ts', function () {
-  let tsResult = gulp.src(['src/**/*.ts', 'typings/**/*.d.ts'])
-    .pipe(tsProject());
+gulp.task("build-ts", function () {
+  const tsResult = gulp.src(["src/**/*.ts", "typings/**/*.d.ts"]).pipe(tsProject());
```
Comment (jp-andre) [inline gulpfile.js:21]: we can probably remove typings from here

File: .prettierignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+/dist
+/src/util/apis/generated
```
Comment (jp-andre) [inline .prettierignore:2]: can we have a newline here? (if prettier accepts)

File: gulpfile.js [hunk 3]
```diff
@@ -1,97 +1,84 @@
...
-const autocompleteTree = require('./scripts/autocomplete-tree');
+const clean = require("gulp-clean");
+const gulp = require("gulp");
+const rimraf = require("rimraf");
+const sourcemaps = require("gulp-sourcemaps");
+const ts = require("gulp-typescript");
+const autocompleteTree = require("./scripts/autocomplete-tree");
 
-let tsProject = ts.createProject('tsconfig.json');
+const tsProject = ts.createProject("tsconfig.json");
```
Comment (jp-andre) [inline gulpfile.js:10]: maybe those can use import syntax? just wondering

File: scripts/all-command-help.js [hunk 4]
```diff
@@ -4,99 +4,99 @@
...
+const os = require("os");
+const path = require("path");
+const util = require("util");
+const Table = require("cli-table3");
+const {
+  getClassHelpText,
+  getOptionsDescription,
+  getPositionalOptionsDescription,
+} = require("../dist/util/commandline/option-decorators");
+const { getOptionsForTwoColumnTableWithNoBorders } = require("../dist/util/interaction/out");
```
Comment (jp-andre) [inline scripts/all-command-help.js:20]: maybe import here as well?

File: scripts/all-command-help.js [hunk 5]
```diff
@@ -4,99 +4,99 @@
...
 function checkStats(dir, fileName, checker) {
   try {
-    let s = fs.statSync(path.join(dir, fileName))
-    return checker(s)
-  }
-  catch (err) {
-    return false
+    let s = fs.statSync(path.join(dir, fileName));
```
Comment (jp-andre) [inline scripts/all-command-help.js:23]: looks like this should be const -- why no linting error?

File: package.json [hunk 6]
```diff
@@ -17,16 +17,18 @@
   },
   "preferGlobal": true,
   "scripts": {
-    "build": "npm run tslint && gulp build:raw",
-    "tslint": "tslint --project .",
-    "test": "npm run tslint && mocha --require ts-node/register ./test/**/*-test.[tj]s",
+    "build": "npm run lint && gulp build:raw",
+    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
```
Comment (jp-andre) [inline package.json:21]: add scripts and bin to that (if possible of course)

**Comment by anywherepilot** [comment]
I'm curious whether you think the changes from for loops to for-of loops are safe. One of them actually caused test failures saying the object wasn't iterable. I don't know how it does pass with a normal for loop. See last commit.

File: .prettierignore (hunk 2 above)
Comment (anywherepilot) [reply .prettierignore:2]: Oh yeah of course. Prettier actually enforces that, but doesn't run on this file type.

Comment (anywherepilot) [reply .prettierignore:2]: Nope it does not enforce that... I must be thinking of some other tool. But I'll add it, no problem.

**Comment by jp-andre** [comment]
> I'm curious whether you think the changes from for loops to for-of loops are safe. One of them actually caused test failures saying the object wasn't iterable. I don't know how it does pass with a normal for loop. See last commit.

Yeah I saw those changes. A quick search shows that the loop will not respect the order of values.
I would rather use indices as there's a risk in changing that.

**Comment by anywherepilot** [comment]
I copied the rules from the common App Center package. We can't import that package here, because it is in the internal feed.

File: gulpfile.js (hunk 1 above)
Comment (anywherepilot) [reply gulpfile.js:21]: Good find. I also removed it from the doc.

File: .prettierignore (hunk 2 above)
Comment (anywherepilot) [reply .prettierignore:2]: Done!

File: gulpfile.js (hunk 3 above)
Comment (anywherepilot) [reply gulpfile.js:10]: I assume you're referring to the `require` statements above. I read up on this syntax a bit. It should work, but doesn't. I'm getting unexpected token errors on the asterisk. Maybe we're not using ES6. I'll leave this for this PR.

File: scripts/all-command-help.js (hunk 4 above)
Comment (anywherepilot) [reply scripts/all-command-help.js:20]: Not touching the import syntax of .js files.

File: scripts/all-command-help.js (hunk 5 above)
Comment (anywherepilot) [reply scripts/all-command-help.js:23]: We never linted JavaScript before. I enabled it and applied fixes.

File: package.json (hunk 6 above)
Comment (anywherepilot) [reply package.json]: Added.

**Pull request merged**